File: JavaScript/5-BinarySearch/Algorithms/1-BinarySearch(Iterative&Recursive)/BinarySearch(Iterative&Recursive)Pro.js
// Binary Search Implementation in JavaScript

// 1. Iterative Binary Search
function iterativeBinarySearch(arr, target) {
  let low = 0;
  let high = arr.length - 1;

  while (low <= high) {
    const mid = Math.floor((low + high) / 2); // More robust against overflow
    const value = arr[mid];

    if (value === target) {
      return mid;
    } else if (value < target) {
      low = mid + 1;
    } else {
      high = mid - 1;
    }
  }
  return -1;
}

// 2. Recursive Binary Search
function recursiveBinarySearch(arr, target, low, high) {
  if (low > high) {
    return -1;
  }

  const mid = Math.floor((low + high) / 2);
  const value = arr[mid];

  if (value === target) {
    return mid;
  } else if (value < target) {
    return recursiveBinarySearch(arr, target, mid + 1, high);
  } else {
    return recursiveBinarySearch(arr, target, low, mid - 1);
  }
}

// 3. Binary Search with Early Exit (Optimization)
function binarySearchEarlyExit(arr, target) {
    let low = 0;
    let high = arr.length - 1;

    while (low <= high) {
        const mid = Math.floor((low + high) / 2);
        const value = arr[mid];

        if (value === target) {
            return mid; // Found, exit early!
        } else if (value < target) {
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }
    return -1; // Not found
}

// 4. Binary Search for First Occurrence
function findFirstOccurrence(arr, target) {
  let low = 0;
  let high = arr.length - 1;
  let result = -1;

  while (low <= high) {
    const mid = Math.floor((low + high) / 2);
    const value = arr[mid];

    if (value === target) {
      result = mid;     // Found, but look for earlier
      high = mid - 1;   // Continue searching left
    } else if (value < target) {
      low = mid + 1;
    } else {
      high = mid - 1;
    }
  }
  return result;
}

// 5. Binary Search for Last Occurrence
function findLastOccurrence(arr, target) {
    let low = 0;
    let high = arr.length - 1;
    let result = -1;

    while (low <= high) {
        const mid = Math.floor((low + high) / 2);
        const value = arr[mid];

        if (value === target) {
            result = mid;       // Found, but look for later
            low = mid + 1;     // Continue searching right
        } else if (value < target) {
            low = mid + 1;
        } else {
            high = mid - 1;
        }
    }
    return result;
}

// Example Usage and Output
const sortedArray = [2, 5, 7, 8, 11, 12, 13, 13, 13, 15, 18, 20];
const targetValue = 13;

console.log("Iterative Binary Search:", iterativeBinarySearch(sortedArray, targetValue)); // Output: 6
console.log("Recursive Binary Search:", recursiveBinarySearch(sortedArray, targetValue, 0, sortedArray.length - 1)); // Output: 6
console.log("Binary Search with Early Exit:", binarySearchEarlyExit(sortedArray, targetValue)); // Output: 6
console.log("First Occurrence of 13:", findFirstOccurrence(sortedArray, targetValue)); // Output: 6
console.log("Last Occurrence of 13:", findLastOccurrence(sortedArray, targetValue));   // Output: 8

// Example with a target not in the array
const notFoundTarget = 99;
console.log("Iterative Binary Search (Not Found):", iterativeBinarySearch(sortedArray, notFoundTarget)); // Output: -1
console.log("Recursive Binary Search (Not Found):", recursiveBinarySearch(sortedArray, notFoundTarget, 0, sortedArray.length - 1)); // Output: -1
console.log("Binary Search with Early Exit (Not Found):", binarySearchEarlyExit(sortedArray, notFoundTarget)); // Output: -1
console.log("First Occurrence of 99:", findFirstOccurrence(sortedArray, notFoundTarget)); // Output: -1
console.log("Last Occurrence of 99:", findLastOccurrence(sortedArray, notFoundTarget));     // Output: -1

// Example with empty array
const emptyArray = [];
const targetInEmpty = 5;
console.log("Iterative Binary Search (Empty Array):", iterativeBinarySearch(emptyArray, targetInEmpty));       // Output: -1
console.log("Recursive Binary Search (Empty Array):", recursiveBinarySearch(emptyArray, targetInEmpty, 0, emptyArray.length-1)); // Output: -1
console.log("Binary Search with Early Exit (Empty Array):", binarySearchEarlyExit(emptyArray, targetInEmpty));       // Output: -1
console.log("First Occurrence (Empty Array):", findFirstOccurrence(emptyArray, targetInEmpty));       // Output: -1
console.log("Last Occurrence (Empty Array):", findLastOccurrence(emptyArray, targetInEmpty));           // Output: -1

// Example with single element array
const singleElementArray = [5];
const targetInSingle = 5;
const targetNotInSingle = 10;
console.log("Iterative Binary Search (Single Element):", iterativeBinarySearch(singleElementArray, targetInSingle));       // Output: 0
console.log("Recursive Binary Search (Single Element):", recursiveBinarySearch(singleElementArray, targetInSingle, 0, singleElementArray.length-1)); // Output: 0
console.log("Binary Search with Early Exit (Single Element):", binarySearchEarlyExit(singleElementArray, targetInSingle));       // Output: 0
console.log("First Occurrence (Single Element):", findFirstOccurrence(singleElementArray, targetInSingle));       // Output: 0
console.log("Last Occurrence (Single Element):", findLastOccurrence(singleElementArray, targetInSingle));           // Output: 0

console.log("Iterative Binary Search (Single Element, Not Found):", iterativeBinarySearch(singleElementArray, targetNotInSingle));       // Output: -1
console.log("Recursive Binary Search (Single Element, Not Found):", recursiveBinarySearch(singleElementArray, targetNotInSingle, 0, singleElementArray.length-1)); // Output: -1
console.log("Binary Search with Early Exit (Single Element, Not Found):", binarySearchEarlyExit(singleElementArray, targetNotInSingle));       // Output: -1
console.log("First Occurrence (Single Element, Not Found):", findFirstOccurrence(singleElementArray, targetNotInSingle));       // Output: -1
console.log("Last Occurrence (Single Element, Not Found):", findLastOccurrence(singleElementArray, targetNotInSingle));           // Output: -1

File: Java/5-BinarySearch/Algorithms/1-BinarySearch(Iterative&Recursive)/BinarySearch(Iterative&Recursive)Pro.java
import java.util.Arrays;
import java.util.InputMismatchException;
import java.util.Scanner;

public class BinarySearch {

    /**
     * Iterative Binary Search Implementation.
     *
     * This method searches for a target value within a sorted array using the iterative approach.
     * It repeatedly divides the search interval in half until the target is found or the interval is empty.
     *
     * @param arr The sorted array to search within.
     * @param target The value to search for.
     * @return The index of the target value if found, otherwise -1.
     *
     * Example Use Case:
     * Searching for a specific product ID in a sorted list of product IDs in an e-commerce application.
     */
    public static int iterativeBinarySearch(int[] arr, int target) {
        int low = 0;
        int high = arr.length - 1;

        while (low <= high) {
            int mid = low + (high - low) / 2; // Prevent potential overflow

            if (arr[mid] == target) {
                return mid; // Target found at the middle
            } else if (arr[mid] < target) {
                low = mid + 1; // Target is in the right half
            } else {
                high = mid - 1; // Target is in the left half
            }
        }

        return -1; // Target not found
    }

    /**
     * Recursive Binary Search Implementation.
     *
     * This method searches for a target value within a sorted array using the recursive approach.
     * It recursively divides the search interval in half until the target is found or the interval is empty.
     *
     * @param arr The sorted array to search within.
     * @param target The value to search for.
     * @param low The starting index of the search interval.
     * @param high The ending index of the search interval.
     * @return The index of the target value if found, otherwise -1.
     *
     * Example Use Case:
     * Implementing a recursive search in a database index to locate a specific record.
     */
    public static int recursiveBinarySearch(int[] arr, int target, int low, int high) {
        if (low > high) {
            return -1; // Base case: Target not found
        }

        int mid = low + (high - low) / 2; // Prevent potential overflow

        if (arr[mid] == target) {
            return mid; // Target found at the middle
        } else if (arr[mid] < target) {
            return recursiveBinarySearch(arr, target, mid + 1, high); // Search in the right half
        } else {
            return recursiveBinarySearch(arr, target, low, mid - 1); // Search in the left half
        }
    }

    /**
     * Binary Search for First Occurrence.
     *
     * This method finds the index of the *first* occurrence of a target value in a sorted array that may contain duplicates.
     *
     * @param arr The sorted array to search within.
     * @param target The value to search for.
     * @return The index of the first occurrence of the target value if found, otherwise -1.
     *
     * Example Use Case:
     * Finding the first occurrence of a specific date in a sorted log file.
     */
    public static int binarySearchFirstOccurrence(int[] arr, int target) {
        int low = 0;
        int high = arr.length - 1;
        int result = -1; // Initialize result to -1 (not found)

        while (low <= high) {
            int mid = low + (high - low) / 2;

            if (arr[mid] == target) {
                result = mid;     // Update result, but continue searching for earlier occurrences
                high = mid - 1;   // Search in the left half
            } else if (arr[mid] < target) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        return result;
    }

    /**
     * Binary Search for Last Occurrence.
     *
     * This method finds the index of the *last* occurrence of a target value in a sorted array that may contain duplicates.
     *
     * @param arr The sorted array to search within.
     * @param target The value to search for.
     * @return The index of the last occurrence of the target value if found, otherwise -1.
     *
     * Example Use Case:
     * Finding the last occurrence of a specific event in a sorted list of events.
     */
    public static int binarySearchLastOccurrence(int[] arr, int target) {
        int low = 0;
        int high = arr.length - 1;
        int result = -1;

        while (low <= high) {
            int mid = low + (high - low) / 2;

            if (arr[mid] == target) {
                result = mid;     // Update result, but continue searching for later occurrences
                low = mid + 1;    // Search in the right half
            } else if (arr[mid] < target) {
                low = mid + 1;
            } else {
                high = mid - 1;
            }
        }
        return result;
    }

    /**
     * Binary Search in Rotated Sorted Array.
     *
     * This method searches for a target value in a sorted array that has been rotated.
     * A rotated sorted array is an array that was originally sorted, but then had some portion of it shifted to the beginning.
     *
     * @param arr The rotated sorted array.
     * @param target The value to search for.
     * @return The index of the target value if found, otherwise -1.
     *
     * Example Use Case:
     * Searching for a specific version number in a deployed application's configuration, where the configuration
     * array might be rotated due to deployment strategies.
     */
    public static int binarySearchRotated(int[] arr, int target) {
        int low = 0;
        int high = arr.length - 1;

        while (low <= high) {
            int mid = low + (high - low) / 2;

            if (arr[mid] == target) {
                return mid;
            }

            if (arr[low] <= arr[mid]) { // Left half is sorted
                if (target >= arr[low] && target < arr[mid]) {
                    high = mid - 1;
                } else {
                    low = mid + 1;
                }
            } else { // Right half is sorted
                if (target > arr[mid] && target <= arr[high]) {
                    low = mid + 1;
                } else {
                    high = mid - 1;
                }
            }
        }
        return -1;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        System.out.println("Binary Search Demo");
        System.out.println("-------------------");

        // Get array size from user
        int arraySize = 0;
        boolean validSize = false;
        while (!validSize) {
            System.out.print("Enter the size of the sorted array: ");
            try {
                arraySize = scanner.nextInt();
                if (arraySize > 0) {
                    validSize = true;
                } else {
                    System.out.println("Array size must be a positive integer.");
                }
            } catch (InputMismatchException e) {
                System.out.println("Invalid input. Please enter an integer.");
                scanner.next(); // Clear the invalid input
            }
        }

        // Create the array
        int[] sortedArray = new int[arraySize];

        // Get array elements from user, ensuring they are sorted
        System.out.println("Enter the elements of the sorted array in ascending order:");
        for (int i = 0; i < arraySize; i++) {
            boolean validElement = false;
            while (!validElement) {
                System.out.print("Element " + (i + 1) + ": ");
                try {
                    int element = scanner.nextInt();
                    if (i > 0 && element < sortedArray[i - 1]) {
                        System.out.println("Element must be greater than or equal to the previous element.");
                    } else {
                        sortedArray[i] = element;
                        validElement = true;
                    }
                } catch (InputMismatchException e) {
                    System.out.println("Invalid input. Please enter an integer.");
                    scanner.next(); // Clear invalid input
                }
            }
        }

        // Get the target value from the user
        System.out.print("Enter the target value to search for: ");
        int target = scanner.nextInt();

        // Perform and display the results of each binary search method
        System.out.println("\nResults:");
        System.out.println("-------------------");

        int iterativeResult = iterativeBinarySearch(sortedArray, target);
        System.out.println("Iterative Binary Search: " + (iterativeResult != -1 ? "Found at index " + iterativeResult : "Not found"));

        int recursiveResult = recursiveBinarySearch(sortedArray, target, 0, sortedArray.length - 1);
        System.out.println("Recursive Binary Search: " + (recursiveResult != -1 ? "Found at index " + recursiveResult : "Not found"));

        int firstOccurrenceResult = binarySearchFirstOccurrence(sortedArray, target);
        System.out.println("First Occurrence: " + (firstOccurrenceResult != -1 ? "Found at index " + firstOccurrenceResult : "Not found"));

        int lastOccurrenceResult = binarySearchLastOccurrence(sortedArray, target);
        System.out.println("Last Occurrence: " + (lastOccurrenceResult != -1 ? "Found at index " + lastOccurrenceResult : "Not found"));

        // Demonstrate on a rotated array
        int[] rotatedArray = {4, 5, 6, 7, 0, 1, 2}; // Example rotated array
        int rotatedResult = binarySearchRotated(rotatedArray, target);
        System.out.println("Rotated Array Search: " + (rotatedResult != -1 ? "Found at index " + rotatedResult : "Not found"));

        scanner.close();
    }
}

